Import implicit methods of CXXRecordDecl.

balazske opened the pull request:
For some (unknown) reason at import of a record definition there may be an existing definition of the same record that does not contain every method. It is possible that automatically generated operators or destructors are missing, for example. The correction does an import of everything if an existing record is found and the existing and to-be-imported is a definition.
(Hopefully the import works in every case for the already existing items without duplicating these.)

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3173,6 +3173,27 @@ TEST_P(
                     .match(ToTU, classTemplateSpecializationDecl()));
 }
 
+TEST_P(ASTImporterTestBase, MergeOfRecordAtImport) {
```
Isn't this an ODR violation? Two TU with different definitions?

balazske replied:
This can be ODR problem but it is not checked yet. It was also not checked before this fix. (To check this the structurally equivalence check must check the methods of a record.)

In an encountered error case there is an existing `CXXRecordDecl` that do not contain destructor, and another one is imported that does contain an automatically generated destructor function. The two classes are in separate header files but otherwise exactly the same source code. It is another question why the original record does not have the destructor but I did not try to figure this out, I do not know at all where to look for related code.

The test is there only to check if the methods are merged. A normal use case is when the first class has no generated destructor (or operators), the second (imported) has these functions, and the source code is the same.

martong replied:
OK.
Only the automatically generated functions were missing? Or was there any other user defined function which was missing?

balazske commented:
There are two variants of a struct that is to be imported: One with automatic function definitions, one without these. The source location is now the same for both (same header, same location). Source code is only a simple struct with field members, nothing else.

The first already existing one (before first import) has declarations only for copy constructor, move constructor, move `operator =`.

The other one (imported for first time) contains the same as above, and these declarations are duplicated here (with a `Prev` pointing to the other one). And there is a destructor, default constructor, copy `operator=` that have an inline body with definition (only one instance).

After the import of the latter into first, result is the same. More imports follow, sometimes with the smaller (without automatic definitions) variant. After the imports the chains of duplicated declarations for copy, move constructor and move `operator=` become longer, up to 4 instances at the end. Destructor, default constructor, copy `operator=` with inline body are not duplicated.

This means that for methods without body redundant redeclaration chains are generated at import. This works for normal member functions too.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2205,8 +2205,12 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
             // function.
-            // FIXME: For C++, we should also merge methods here.
-            return Importer.MapImported(D, FoundDef);
+            Importer.MapImported(D, FoundDef);
```
The enclosing condition is true if the class to be imported is either has a name and incomplete or it is anonymous and complete. I don't see how this is satisfied when the class to be imported has a name and is complete (the class in the test seems like having those properties).

balazske replied:
The class to be imported is complete and is not anonymous and the `FoundDef` is not anonymous too. The condition is true then.

martong replied:
Oh man, ok. This condition is hard to read.

martong commented:
So you say that by merging the redecl chains will grow? I think that is ok.
Or will there be redundant redecl chains for the same function? That is not ok. Could you please add more asserts/expects about the redecl chain(s)?

balazske commented:
The AST looks like this, after imports:
```
CXXRecordDecl 0x1dfc270 <Xxx.h:121:3, line:128:3> line:121:10 referenced struct Xxx definition
|-DefinitionData standard_layout can_const_default_init
| |-DefaultConstructor exists non_trivial user_provided
| |-CopyConstructor simple non_trivial has_const_param needs_overload_resolution implicit_has_const_param
| |-MoveConstructor exists simple non_trivial needs_overload_resolution
| |-CopyAssignment non_trivial has_const_param implicit_has_const_param
| |-MoveAssignment exists simple non_trivial needs_overload_resolution
| `-Destructor simple non_trivial
|-CXXConstructorDecl 0x1e26430 <col:10> col:10 implicit Xxx 'void (const Xxx &)' inline default noexcept-unevaluated 0x1e26430
| `-ParmVarDecl 0x1e26560 <col:10> col:10 'const Xxx &'
|-CXXConstructorDecl 0x1e26d80 <col:10> col:10 implicit Xxx 'void (Xxx &&)' inline default noexcept-unevaluated 0x1e26d80
| `-ParmVarDecl 0x1e26eb0 <col:10> col:10 'Xxx &&'
|-CXXMethodDecl 0x1e26f48 <col:10, <invalid sloc>> col:10 implicit operator= 'Xxx &(Xxx &&)' inline default noexcept-unevaluated 0x1e26f48
| `-ParmVarDecl 0x1e27070 <col:10> col:10 'Xxx &&'
|-CXXConstructorDecl 0x35ce6e8 prev 0x1e26430 <col:10> col:10 implicit Xxx 'void (const Xxx &)' inline noexcept-unevaluated 0x1e26430
| `-ParmVarDecl 0x35ce688 <col:10> col:10 'const Xxx &'
|-CXXConstructorDecl 0x35ce800 prev 0x1e26d80 <col:10> col:10 implicit Xxx 'void (Xxx &&)' inline noexcept-unevaluated 0x1e26d80
| `-ParmVarDecl 0x35ce7a0 <col:10> col:10 'Xxx &&'
|-CXXMethodDecl 0x35ce918 prev 0x1e26f48 <col:10, <invalid sloc>> col:10 implicit operator= 'Xxx &(Xxx &&)' inline noexcept-unevaluated 0x1e26f48
| `-ParmVarDecl 0x35ce8b8 <col:10> col:10 'Xxx &&'
|-CXXDestructorDecl 0x35ce9c0 <col:10> col:10 implicit used ~Xxx 'void () noexcept' inline
| `-CompoundStmt 0x35cea70 <col:10>
|-CXXConstructorDecl 0x35cea80 <col:10> col:10 implicit used Xxx 'void () noexcept' inline
| |-CXXCtorInitializer Field 0x1e25588 'fff2' 'std::shared_ptr<Yyy>':'std::shared_ptr<Yyy>'
| | `-CXXConstructExpr 0x4053168 <col:10> 'std::shared_ptr<Yyy>':'std::shared_ptr<Yyy>' 'void () noexcept'
| `-CompoundStmt 0x40531c8 <col:10>
|-CXXMethodDecl 0x40532b8 <col:10> col:10 implicit used operator= 'Xxx &(const Xxx &) noexcept' inline
| |-ParmVarDecl 0x4053258 <col:10> col:10 used 'const Xxx &'
| `-CompoundStmt 0x4054350 <col:10>
|   |-BinaryOperator 0x4053590 <col:10> 'uint8_t':'unsigned char' lvalue '='
|   | |-MemberExpr 0x40534e0 <col:10> 'uint8_t':'unsigned char' lvalue ->fff 0x1dfc498
|   | | `-CXXThisExpr 0x40534c8 <col:10> 'Xxx *' this
|   | `-ImplicitCastExpr 0x4053578 <col:10> 'uint8_t':'unsigned char' <LValueToRValue>
|   |   `-MemberExpr 0x4053540 <col:10> 'const uint8_t':'const unsigned char' lvalue .fff 0x1dfc498
|   |     `-DeclRefExpr 0x4053518 <col:10> 'const Xxx' lvalue ParmVar 0x4053258 '' 'const Xxx &'
|   |-CXXMemberCallExpr 0x40542d0 <col:10> 'std::shared_ptr<Yyy>' lvalue
|   | |-MemberExpr 0x4054220 <col:10> '<bound member function type>' .operator= 0x40536a8
|   | | `-MemberExpr 0x40535d0 <col:10> 'std::shared_ptr<Yyy>':'std::shared_ptr<Yyy>' lvalue ->fff2 0x1e25588
|   | |   `-CXXThisExpr 0x40535b8 <col:10> 'Xxx *' this
|   | `-MemberExpr 0x4054298 <col:10> 'const std::shared_ptr<Yyy>':'const std::shared_ptr<Yyy>' lvalue .fff2 0x1e25588
|   |   `-DeclRefExpr 0x4054270 <col:10> 'const Xxx' lvalue ParmVar 0x4053258 '' 'const Xxx &'
|   `-ReturnStmt 0x4054338 <col:10>
|     `-UnaryOperator 0x4054318 <col:10> 'Xxx' lvalue prefix '*'
|       `-CXXThisExpr 0x4054300 <col:10> 'Xxx *' this
|-CXXConstructorDecl 0x5209de0 prev 0x35ce6e8 <col:10> col:10 implicit Xxx 'void (const Xxx &)' inline noexcept-unevaluated 0x1e26430
| `-ParmVarDecl 0x5209d80 <col:10> col:10 'const Xxx &'
|-CXXConstructorDecl 0x5209ef8 prev 0x35ce800 <col:10> col:10 implicit Xxx 'void (Xxx &&)' inline noexcept-unevaluated 0x1e26d80
| `-ParmVarDecl 0x5209e98 <col:10> col:10 'Xxx &&'
|-CXXMethodDecl 0x520a010 prev 0x35ce918 <col:10, <invalid sloc>> col:10 implicit operator= 'Xxx &(Xxx &&)' inline noexcept-unevaluated 0x1e26f48
| `-ParmVarDecl 0x5209fb0 <col:10> col:10 'Xxx &&'
|-CXXRecordDecl 0x1dfc398 <col:3, col:10> col:10 implicit struct Xxx
|-FieldDecl 0x1dfc498 <line:124:5, col:13> col:13 fff 'uint8_t':'unsigned char'
|-FieldDecl 0x1e25588 <line:125:5, col:33> col:33 fff2 'std::shared_ptr<Yyy>':'std::shared_ptr<Yyy>'
|-CXXConstructorDecl 0x604bfa0 prev 0x5209de0 <line:121:10> col:10 implicit Xxx 'void (const Xxx &)' inline noexcept-unevaluated 0x1e26430
| `-ParmVarDecl 0x604bf40 <col:10> col:10 'const Xxx &'
|-CXXConstructorDecl 0x604c0b8 prev 0x5209ef8 <col:10> col:10 implicit Xxx 'void (Xxx &&)' inline noexcept-unevaluated 0x1e26d80
| `-ParmVarDecl 0x604c058 <col:10> col:10 'Xxx &&'
`-CXXMethodDecl 0x604c1d0 prev 0x520a010 <col:10, <invalid sloc>> col:10 implicit operator= 'Xxx &(Xxx &&)' inline noexcept-unevaluated 0x1e26f48
  `-ParmVarDecl 0x604c170 <col:10> col:10 'Xxx &&'
```
If there is an other member function declaration (no body) the dump fails with "unsortable locations found" after this function is printed. The test on other projects produced failures too.

martong commented:
This is how it looks before or after your fix?

balazske commented:
After the fix. Before the fix this is present (and after the import this remains because the already existing definition is used):
```
CXXRecordDecl 0x1dfc270 <Xxx.h:121:3, line:128:3> line:121:10 referenced struct Xxx definition
|-DefinitionData standard_layout can_const_default_init
| |-DefaultConstructor exists non_trivial user_provided
| |-CopyConstructor simple non_trivial has_const_param needs_overload_resolution implicit_has_const_param
| |-MoveConstructor exists simple non_trivial needs_overload_resolution
| |-CopyAssignment non_trivial has_const_param implicit_has_const_param
| |-MoveAssignment exists simple non_trivial needs_overload_resolution
| `-Destructor simple non_trivial
|-CXXConstructorDecl 0x1e26430 <col:10> col:10 implicit Xxx 'void (const Xxx &)' inline default noexcept-unevaluated 0x1e26430
| `-ParmVarDecl 0x1e26560 <col:10> col:10 'const Xxx &'
|-CXXConstructorDecl 0x1e26d80 <col:10> col:10 implicit Xxx 'void (Xxx &&)' inline default noexcept-unevaluated 0x1e26d80
| `-ParmVarDecl 0x1e26eb0 <col:10> col:10 'Xxx &&'
|-CXXMethodDecl 0x1e26f48 <col:10, <invalid sloc>> col:10 implicit operator= 'Xxx &(Xxx &&)' inline default noexcept-unevaluated 0x1e26f48
| `-ParmVarDecl 0x1e27070 <col:10> col:10 'Xxx &&'
|-CXXRecordDecl 0x1dfc398 <col:3, col:10> col:10 implicit struct Xxx
|-FieldDecl 0x1dfc498 <line:124:5, col:13> col:13 fff 'uint8_t':'unsigned char'
|-FieldDecl 0x1e25588 <line:125:5, col:33> col:33 fff2 'std::shared_ptr<Yyy>':'std::shared_ptr<Yyy>'
```

martong commented:
Ok, now I see what is happening.

I understand that technically merging the two definitions is possible and the redecl chain of each method will be handled properly (because we handle the redecl chain in case of any functions properly). 

But, this kind of merging could result a very obfuscated AST for classes and methods.
This is why I think first we should understand how is that possible that one definition of a class contains different methods than an other (because that is the reason which forces us into this fix).

Once we know that then we can consider this as a fix or we could come up with other solutions.
Also, in case of free functions we build up a similar long redecl chain, but there we don't have any other options. If it turns out that this merge is the only solution then we should implement the squashing of redundant function declarations as a next step.

balazske commented:
Code was now updated, now we get no redecl chain. Still improvements are needed for friend and friend-template cases. Some of the tests fails because functionality is changed:
```
ImportFunctions.ImportPrototypeThenProtoAndDefinition
ImportFunctions.ImportPrototypeThenPrototype
ImportFunctions.ImportPrototypes
ImportFunctions.PrototypeAfterPrototype
```

martong commented:
Building the whole redecl chain in case of free functions was deliberate. We know that once we should squash the redundant prototypes. But that didn't look very important and I think that should be in a totally separate pull request with several number of different tests to verify and document what exactly we do with redundant prototypes.
Now with this PR you are trying to coalesce two improvements and I am still not convinced that squashing the prototypes is the fix for the missing destructor. As I mentioned previously I consider finding the cause of the different class definitions more important than having a quick fix.

martong commented:
> For some (unknown) reason at import of a record definition there may be an existing definition of the same record that does not contain every method.

I think we should find that reason.

balazske commented:
The problem is that the default functions are lazy-created. If an existing class do not contain and the imported version of it does contain these, they will not be imported and not created. To create these, a `Sema` instance is needed that is not available in `ASTImporter`.

balazske committed:
[ASTImporter] Import implicit functions.

If a CXXRecordDecl with definition is imported and there is an existing
definition, update the definition with implicit functions from the
imported one.
The implicit functions may not be present in the existing definition,
if the existing code did not make use of them. But the imported code may
have references to these functions.

balazske commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3352,5 +3451,9 @@ INSTANTIATE_TEST_CASE_P(
 
+INSTANTIATE_TEST_CASE_P(
+    ParameterizedTests, ImportImplicitMethods,
+    ::testing::Values(ArgVector(), ArgVector{"-fdelayed-template-parsing"}),);
```
Is delayed-template-parsing needed (no templates in the test code)?

martong replied:
I think it is better to leave it there, to guarantee that we avoid future errors if we add a template test to `ImportImplicitMethods`. Also, soon we will execute all tests with the same switch sets, see https://reviews.llvm.org/D47367

balazske commented:
Code update: Now the "implicit" functions are imported only, when importing a record definition into an existing definition.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2205,8 +2216,17 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
-            return Importer.MapImported(D, FoundDef);
+            // FIXME: Check for ODR.
+            Importer.MapImported(D, FoundDef);
+            if (auto *DCXX = dyn_cast<CXXRecordDecl>(D)) {
```
`const auto *DCXX`

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3173,6 +3173,105 @@ TEST_P(
+  auto Code = R"(
+      struct A { A() { int x; } };
+      )";
+  test(cxxConstructorDecl(isDefaultConstructor()), Code, true);
```
`/*CheckNoImport=*/true` makes it easier to read.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -1211,6 +1212,16 @@ void ASTNodeImporter::ImportDeclContext(DeclContext *FromDC, bool ForceImport,
+      "Import implicit methods to or from non-definition");
+  
+  for (CXXMethodDecl *FromM : From->methods())
+    if (FromM->isImplicit())
```
So `isImplicit` is true for all non-user-defined methods? And only non-user-defined methods have `isImplicit() == true`?

balazske replied:
It looks like that `isImplicit` is set for everything that is compiler-generated, therefore it should be correct to use here.

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3173,6 +3173,105 @@ TEST_P(
+      void f() {
+        A a;
+        A a1(a);
+        A a2(fa());
```
`A a2(A());` would result the same, wouldn't it? The `A()` expression has a `prvalue` value category, thus it should involve generating the implicit move ctor. And then `fa()` is superfluous.

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3173,6 +3173,105 @@ TEST_P(
+
+    ASSERT_EQ(DeclCounter<CXXMethodDecl>().match(ToClass, MethodMatcher), 0);
+
+    Decl *ImportedClass;
```
` = nullptr;` ? Just to make it initialized always.

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3173,6 +3173,105 @@ TEST_P(
+
+  template<typename MatcherType>
+  void test(const MatcherType &MethodMatcher,
+      const char *Code = DefaultCode, bool CheckNoImport = false) {
```
Why don't we just pass the `ExpectedCount` as a parameter? Default could be `1`.
`CheckNoImport` is an opaque name for me, I had to understand the whole `test` code before knowing its real meaning.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -1211,6 +1212,16 @@ void ASTNodeImporter::ImportDeclContext(DeclContext *FromDC, bool ForceImport,
+  assert(From->isCompleteDefinition() && To->getDefinition() == To &&
+      "Import implicit methods to or from non-definition");
+  
+  for (CXXMethodDecl *FromM : From->methods())
```
`const` ?

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3173,6 +3173,105 @@ TEST_P(
+
+    ASSERT_EQ(DeclCounter<CXXMethodDecl>().match(ToClass, MethodMatcher), 1);
+
+    {
```
Do we need the blocks? Is there a name collision, or you just want to express that `Method` is not used in the final expectations?

balazske replied:
This is a kind of sub-operation with local variables for it. (Other tests have same constructs, probably because name conflicts). The `Method` is anyway not needed later, it points to a removed thing that should not be used any more.

martong reviewed (commented):
I like the new approach, nice work. Found only some minor things.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2205,8 +2216,17 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
             // function.
-            // FIXME: For C++, we should also merge methods here.
-            return Importer.MapImported(D, FoundDef);
+            // FIXME: Check for ODR.
```
Why do we need this FIXME? Didn't we check already for ODR? There is a structural match at this point, so we know exactly that this is a similar definition with the exact same field members.

balazske replied:
Methods can be different in the two record definitions (theoretically).

martong replied:
The ODR check of the methods is happening when we call the import on that method (structural eq check does that), so I still don't think this `FIXME` is valid.

balazske replied:
The bad case can be if there is a class:
```
struct X { void f1(); };
```
And such is imported:
```
struct X { void f2(int); };
```
If not the merge (X will contain f1 and f2) is the correct way of importing this there could be a warning for this case. (If the ASTImporter is supposed to work only with source code of a single valid program probably no such case happens.)

balazske replied:
Probably the FIXME can be changed to "check for same methods".

martong replied:
Ok, but that bad case should be checked by the structural equivalence before we get into this block. So that should be a FIXME for the structural eq,  check the methods in structural eq, isn't it?

martong reviewed (approved):
OK, but I still think the FIXME is not relevant and should be removed.

balazske commented:
FIXME updated, tests updated for last ctu-clang6.

balazske commented:
Upstream revision:  https://reviews.llvm.org/D49245

balazske commented:
Backport from upstream: #438
